### PR TITLE
Feature#7/list branch create and edit

### LINKED ISSUE #7: List branch, create and edit
# REQUIREMENTS
This screen should list all the branches registered in the app, and the user can search branch or edit from the list view,
PD: The edit action is a redirect to another view of the app

# DESIGN
Provide references to images of the design, how it looks like now, and how it should look like.

# Acceptance Criteria
List all required acceptance criteria
- [ ] The list must contain a pagination
- [ ] The list must contain a searcher
- [ ] The user admin can edit but cannot delete a branch from the list
- [ ] The search engine should work for all branch fields
- [ ] The search field accepts 400 characters

### PR DESCRIPTION
**Pull request by MiguelOcegueraM**
Created the branches manager 
 
 
## Description: Why did you change/implement that?
This tool helps to add branches in the business

## Testing: Help me how can I test or look at the changes?
image
image
image

### CODE REVIEW

File: reloj_checador_team03/app/helpers/attendances_helper.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module AttendancesHelper
```
Comment (OchoaHub) [inline reloj_checador_team03/app/helpers/attendances_helper.rb:1]: If you don't need helpers, delete it. This was recommended to me

**Comment by jnegrete15** [comment]
Goog job, the web design seems great!

File: reloj_checador_team03/app/helpers/branches_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module BranchesHelper
```
Comment (OchoaHub) [inline reloj_checador_team03/app/helpers/branches_helper.rb:1]: By Yil Kanda that suggestion

**Review by OchoaHub — APPROVED** [review]

File: reloj_checador_team03/app/helpers/branches_helper.rb (hunk 2 above)
Comment (MiguelOcegueraM) [reply reloj_checador_team03/app/helpers/branches_helper.rb:1]: thanks for the suggestion

File: reloj_checador_team03/app/helpers/attendances_helper.rb (hunk 1 above)
Comment (MiguelOcegueraM) [reply reloj_checador_team03/app/helpers/attendances_helper.rb:1]: thanks

File: reloj_checador_team03/Gemfile [hunk 3]
```diff
@@ -16,6 +16,7 @@ gem 'webpacker', '~> 5.0'
 # Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks
 gem 'turbolinks', '~> 5'
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
+gem 'wdm', '>= 0.1.0' if Gem.win_platform?
```
Comment (EduardoGHdez) [inline reloj_checador_team03/Gemfile:19]: This gem seems to be more into the `development` bucket.

Right below in this file 😄 

I'd also suggest committing this alone, and describe in your commit why are you adding this dependency to the project, since this does not look like something related to kaminari nor the `employees-feature`

File: reloj_checador_team03/app/assets/stylesheets/attendances.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Attendances controller here.
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/assets/stylesheets/attendances.scss:1]: If it's not necessary, there is no need on adding it in the project

Try to only add what is necessary in order to complete the feature you are describing in you commit message.

I think is okay for now. I mean, this is a toy-project
But take it into consideration for future contributions 😄

File: reloj_checador_team03/app/assets/stylesheets/employees.scss [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
+
+.table-thead {
+    background-color: #7AC143;
+}
+
+.table-thead-th{
+    color: #fff;
+}
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/assets/stylesheets/employees.scss:11]: As you can see here, GitHub helps us to notice the lack of EOF in this file.
Here is a blog that describes better why is it needed:  https://thoughtbot.com/blog/no-newline-at-end-of-file

Sharing the blog in order to save some time trying to describe it by my self XD

File: reloj_checador_team03/app/assets/stylesheets/fontawesome.scss [hunk 6]
```diff
@@ -0,0 +1,3 @@
+.fa-edit, .fa-user-times{
+    color: #7AC143;
+}
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/assets/stylesheets/fontawesome.scss:3]: Same as above

File: reloj_checador_team03/app/controllers/employees_controller.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
+class EmployeesController < ApplicationController
+  before_action :set_employee, only: %i[ show edit update destroy ]
+
+  # GET /employees or /employees.json
+  def index
+    @employees = Employee.all.page(params[:page])
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/controllers/employees_controller.rb:6]: I like you are using pagination from the beginning 🔥

File: reloj_checador_team03/app/models/employee.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class Employee < ApplicationRecord
+    def self.search(text)
+        if text.blank?  # blank? covers both nil and empty string
+          all
+        else
+          where('name LIKE ? or last_name LIKE ? or email LIKE ?', "%#{text}%", "%#{text}%", "%#{text}%")
+        end
+    end
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/models/employee.rb:8]: Nice! 🔥

File: reloj_checador_team03/app/views/attendances/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,17 @@
+
+<%= simple_form_for(@attendance) do |f| %>
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/views/attendances/_form.html.erb:2]: And also using simple-form 👍🏽 
Nice!

File: reloj_checador_team03/config/database.yml [hunk 10]
```diff
@@ -24,8 +24,8 @@ default: &default
 development:
   <<: *default
   database: reloj_checador_team03_development
-  username: miguel
-  password: tscom2020
+  username: fego
+  password: kurosu
   host: localhost
   port: 5432
```
Comment (EduardoGHdez) [inline reloj_checador_team03/config/database.yml]: It's a good practice to not track this file since we are able to config as we want locally.

I'd suggest moving this file to `database.yml.example` as a placeholder of what is needed in order to make the app run locally.

File: reloj_checador_team03/test/system/employees_test.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+  test "destroying a Employee" do
+    visit employees_url
+    page.accept_confirm do
+      click_on "Destroy", match: :first
+    end
+
+    assert_text "Employee was successfully destroyed"
+  end
+end
```
Comment (EduardoGHdez) [inline reloj_checador_team03/test/system/employees_test.rb:51]: You are covering so much for a single commit

Considering splitting in more atomic tasks for the next time 😄 
I know are just CRUD implementations, but worth splitting it in order to make it easier to review

Also, saying in the commit-message the "employees feature is done" is not very descriptive.

I think this is something we can improve the next time.

But have to be honest, I'm super impressed with the code-quality in this first commit!

For more info about how to write better commits, read this blog, I'm pretty sure you'd enjoy it: https://chris.beams.io/posts/git-commit/

Comment (EduardoGHdez) [inline reloj_checador_team03/test/system/employees_test.rb:51]: 🔥 

NICE!!

Comment (EduardoGHdez) [reply reloj_checador_team03/test/system/employees_test.rb:51]: Love it love it love it ❤️

File: reloj_checador_team03/app/views/branches/index.html.erb [hunk 12]
```diff
@@ -1,12 +1,31 @@
...
 
-<table>
-  <thead>
+<div class="container-content">
+
+  <div class="d-flex justify-content-between bd-highlightm mb-3">
+    <div class="d-flex justify-content-start">
+     <%= #form_tag(branch_search_path, method: "post", remote: true, authenticity_token: true, role: "form",class: "form") do %>
+     <%= #text_field_tag('search', nil, placeholder: 'Enter search term...', class: 'form-control') %>
+     <% #end %>
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/views/branches/index.html.erb]: Is this still needed?

File: reloj_checador_team03/config/database.yml [hunk 13]
```diff
@@ -24,8 +24,8 @@ default: &default
 development:
   <<: *default
   database: reloj_checador_team03_development
-  username: fego
-  password: kurosu
+  username: mike
+  password: tscom2020
   host: localhost
   port: 5432
```
Comment (EduardoGHdez) [inline reloj_checador_team03/config/database.yml]: Mentioned previously, worth stop tracking this file and replace it with a `.example` one

File: reloj_checador_team03/config/initializers/content_security_policy.rb [hunk 14]
```diff
@@ -28,3 +28,7 @@
 # For further information see the following documentation:
 # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only
 # Rails.application.config.content_security_policy_report_only = true
+
+
+# config/initializers/content_security_policy.rb
+#policy.connect_src :self, :https, "http://localhost:3035", "ws://localhost:3035" if Rails.env.development?
```
Comment (EduardoGHdez) [inline reloj_checador_team03/config/initializers/content_security_policy.rb:34]: ❓

File: reloj_checador_team03/config/webpack/environment.js [hunk 15]
```diff
@@ -1,14 +1,3 @@
...
-environment.plugins.prepend('Provide',
-    new webpack.ProvidePlugin({
-        $: 'jquery/src/jquery',
-        jQuery: 'jquery/src/jquery',
-        Popper: ['popper.js', 'default']
-    })
-)
-
-
-module.exports = environment
```
Comment (EduardoGHdez) [inline reloj_checador_team03/config/webpack/environment.js]: This would still be needed if you depend on bootstrap 4 at least 🤔 

Why are we removing it?

File: reloj_checador_team03/config/webpack/environment.js [hunk 16]
```diff
@@ -1,3 +1,14 @@
...
+environment.plugins.prepend('Provide',
+    new webpack.ProvidePlugin({
+        $: 'jquery/src/jquery',
+        jQuery: 'jquery/src/jquery',
+        Popper: ['popper.js', 'default']
+    })
+)
+
+
+module.exports = environment
```
Comment (EduardoGHdez) [inline reloj_checador_team03/config/webpack/environment.js]: 😆

File: reloj_checador_team03/app/views/branches/_form.html.erb [hunk 17]
```diff
@@ -1,7 +1,7 @@
 <div class="modal-dialog">
 <% @branch= Branch.new if @branch.nil?  %>
```
Comment (EduardoGHdez) [inline reloj_checador_team03/app/views/branches/_form.html.erb:2]: Not part of this change, but try to have as little as possible this kind of logic from the views
You can instantiate this from the `Branches#new` definition 😄

File: reloj_checador_team03/test/controllers/tmp/rubycritic/assets/javascripts/application.js [hunk 18]
```diff
@@ -0,0 +1,281 @@
+prettyPrint();
```
Comment (EduardoGHdez) [inline reloj_checador_team03/test/controllers/tmp/rubycritic/assets/javascripts/application.js:1]: HERE BE DRAGONS 🐉 

There is no need on tracking these`tmp` files

You can always run Rubycritic in your CI pipeline (it could be GitHub Actions)

**Comment by EduardoGHdez** [comment]
Looks good overall!

For real, what a nice job!

Consider my suggestion in order to continue improving! 🔥 
Congratulations 😄

**Review by EdsonMorfin — APPROVED** [review]
excellent

**Comment by fego-dev** [comment]
Good Job

**Comment by fego-dev** [comment]
Good job

**Review by jnegrete15 — APPROVED** [review]

**Review by fego-dev — APPROVED** [review]
Good job
